src/lib/krb5/krb/s4u_creds.c

```diff
@@ -405,7 +405,8 @@ convert_to_enterprise(krb5_context context, krb5_principal princ,
     char *str;
 
     *eprinc_out = NULL;
-    code = krb5_unparse_name(context, princ, &str);
+    code = krb5_unparse_name_flags(context, princ,
+                                   KRB5_PRINCIPAL_UNPARSE_DISPLAY, &str);
```

Comment:
What's this change about?  If we're going to immediately re-parse the name, won't the quoted special characters be de-quoted at that point?

Reply:
Not quite, it causes the reparse to escape the first at, resulting 3 at's at the end, causing kinit -E followed by s4u2self to fail.

Reply:
So in this case, in_creds->server is already an enterprise principal?

Reply:
Yes, but it may be an enterprise without inner at, so by unparsing and reparsing we get the right result.

Reply:
You're saying if we have an enterprise principal with data[0] = 'abc' and a realm set, we want to translate that to data[0] = abc@REALM?

Reply:
Yes, that our best shot to get a routable server name (the name-type matters less, what matters is the inner at).

Reply:
I think we're introducing a bug here if princ is a regular principal with an intentional `@` (not intended as either a realm or enterprise domain separator) in the first data component.  But nobody does that as it would just be inviting trouble, so it's probably not a bug with real consequences.

I thought about making krb5_parse_name_flags() treat `\@` as the first at for an enterprise name, but Heimdal doesn't do that, so it doesn't seem like a good idea.

Reply:
I think if it's only one at sign, then the name-type will just change to enterprise but the name will stay as is.